processing/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -205,14 +213,20 @@ public boolean hasNext()
                       currentBatch = queue.take();
                     }
                   }
-                  if (currentBatch == null) {
-                    throw new QueryTimeoutException();
-                  }
 
-                  if (cancellationGizmo.isCancelled()) {
+                  if (cancellationGizmo.isCancelled()) { // preffering to throw any cancellation exception, if any
                     throw cancellationGizmo.getRuntimeException();
                   }
 
+                  if (currentBatch == null) {
+                    LOG.info("Query timed out while merging results from data nodes. "
+                             + "Current merge state is [%s] and time spent in current state is [%,d]ms.",
+                             mergeCombineMetricsAccumulator.getState(),
+                             mergeCombineMetricsAccumulator.timeSpentInCurrentStateNanos()
+                    );
```

Comment:
info doesn't really seem right for this, should this information just decorate the timeout exception instead since that is also going to be logged?